### DISCUSSION TITLE
How to make content more classroom proof?

### DISCUSSION THREAD

**Discussion by Felienne** [hedyorg/hedy#3930] [Community]
For a while now I have not been super happy about the content it its current form. This will be a long discussion and has a high "note to self" percentage but I am still very curious what others feel (esp teachers!).

### The current issues

When teaching, I note a number of things (most important issues first).

**What to do?**
It is unclear what to do, when an assignment is ready or complete. For example in the story in level 5:

<img width="691" border="2px" alt="image"  src="https://user-images.githubusercontent.com/1003685/212007955-211fea7f-2475-4a30-bcd0-b66a093b072b.png">

My goal with the exercise (which happens with some students, but surely not all!) is to change the story into something different, with their own characters and endings. Some students just copy the example and say: look a story with 2 endings, and only change one line (see below).

**Copying examples**
A related but different issue is that you can just copy the example, easily enough with the button, and be done. In the above exercises, the description of what to do is exactly the same as the example program and you have done all it says in the exercise.

I do want to stress that it is a different problem, because in the calculator assignment in level 6, shown below, I have started to telll my students to add at least 3 more, different calculations and add one `ask`, which is a useful and fun addition to the starting code.

<img width="1418" alt="image" src="https://user-images.githubusercontent.com/1003685/212008728-ec3c44b5-297d-472b-8d48-685f4c414230.png">

So just the code is ok, I think, as long as we clearly ask an extension to it, it is about the combination of the two I think.

**When are students 'ready'?**
Let's look at the level 5 story again. A third and again slightly different issue is that it is not clear when kids are done. There are instructions to make a story with two endings... is that enough? How many lines are required? We typically add some extra requirements in class (25 lines, 3 variables etc) but that content is now not on the platform.

**Skipping Intro and What's Next?**
Despite me making it an explicit goal in my slides, students do not read and especially skip the Introduction and What's Next tab.

### The root cause

I also want to dive a bit onto how the content came to be as an interesting reflection. I now realize that when I was building Hedy, I was *not* thinking of classroom use. Despite having the explicit goal of using it in my class, when designing the initial content, I was thinking of a kid at home, more problematic, I was thinking of me as a 10 year old. Not so very explicitly of course, but I was assuming a kids that wants to program, and upon being presented with a code snippet wants to pick it apart, change it and see how it works. And while I of course knew not all kids were like that, it was still something I subconsciously designed for

And then, of course, we launched in the pandemic, during which my stereotype was largely true, because our usage then was (what I remember, anecdotically from emails and tweets) parents with kids stuck at home; parents who would say "let's now try this" or "Do that". This was before we had teacher accounts and classes etc, there wasn't really any classroom teaching at that point in time (not by myself too, or I would have seen these issues sooner).

I think all of the Hedy team (me mostly!) really needs to make the shift towards designing for kids (and teachers) who don't necessarily want to learn programming, and we need to lure them in with fun but clearly defined steps.

### Why is this apparent now?

I am not sure why it is becoming more and more visible but I think a reason is that in early versions of Hedy, there were a lot of errors still! There was a time where we did not even test the code snippets and those could (and did) have errors! If the main issue in a class room is errors, then you do not have time really to considers lessons so deeply (is there a lesson here for all of CS1 ed...?)

**Comment by Felienne** [comment]
I will document my changes here so it is easier for people to follow along and comment, and easier for me to do partial work. This is in order of levels, so not chronological!

#4101 (by @MarleenGilsing):
* Makes dice, dishes and rock for level 5 more explicit

#3935:
* removed the turtle from level 7 (with one line of `repeat` there is not a lot to be done and it only raises questions)
* renamed "challenge" to "exercise"

#3942:
* Makes the exercises in level 8 more explicit, referring explicitly to example code in multiple places
* Removes Press it adventure from level 8

#3953:
* Makes the exercises in level 9 more explicit, referring explicitly to example code in multiple places
* Adds Turtle adventure in level 9, removes second example for Story

#3994
* Makes the exercises in level 10 more explicit, changes the order of adventures
* Removes long song Old MacDonald example (even though I love it!)

#3997
* Makes the exercises in level 11 more explicit, changes the order of adventures
* Adds calculator in level 11
* Removes long song Haunted House adventure

#4097
* Makes the exercises in level 12 and level 13 more explicit (partly rework of #4034)
* Changes the example code of tic tac toe using new language features (added in #4047 and #4106)

**Comment by MarleenGilsing** [comment]
Can we add awards/accomplishments/turn the adventure a different color/etc if a student runs a code in an adventure? Then we could make the exercises super explicit (like: 22 lines of code, at least 3 if statements, cannot contain the same lines of code as the example code)? Or would that be really hard to do?
 
Because in that case, we could turn each challenge into a very specific set of criteria. And if the system can check if the students have met the criteria in an adventure, it would be really convenient for a teacher to track the students' progress. And we could display that in a nice graph/table at the classroom page.

**Reply by MarleenGilsing** [reply]
I also still really like the idea of having an outputscreen that runs a code without showing the input and the students have to try to recreate the program.

**Reply by Felienne** [reply]
> Can we add awards/accomplishments/turn the adventure a different color/etc if a student runs a code in an adventure? Then we could make the exercises super explicit (like: 22 lines of code, at least 3 if statements, cannot contain the same lines of code as the example code)? Or would that be really hard to do?

I am not 100% sure I understand the suggestion, do you propose to turn the tabs a different color when they have an explicit assignment? My goal for now is to have such a exercise in all adventure.

> Because in that case, we could turn each challenge into a very specific set of criteria. And if the system can check if the students have met the criteria in an adventure, it would be really convenient for a teacher to track the students' progress. And we could display that in a nice graph/table at the classroom page.

This can be done, but there are some ery interesting educational considerations here, one day I will write a longer post on these too! But for now it suffices to say it would be a lot of work to make such a check go automatically!

**Reply by Felienne** [reply]
> I also still really like the idea of having an outputscreen that runs a code without showing the input and the students have to try to recreate the program.

Yes, me too!! Maybe we (by which I mean: you :) ) can add a few of these in one adventure?

**Comment by Felienne** [comment]
I am continuing work on the adventures again for the higher levels. One observation I am making is about very long code snippets:

![image](https://user-images.githubusercontent.com/1003685/216266951-ae4db4a6-ce6c-4396-87f5-73696dbd244a.png)

I don't think they are very helpful. They seem to be quite overwhelming even for good students and don't necessarily lead to action (change a few things) apart from for the really good students. This might be different if the exercises are more clear but I think with larger programs for now it is better to have them build them step by step, so I am removing a few larger programs for now.

**Reply by MarleenGilsing** [reply]
Maybe we can add these longer codes to the explore page, so they can function as an inspiration/example tehy can play with, instead of scaring off the students.

**Comment by MarleenGilsing** [comment]
Another hersenspinsel about helping teachers monitor progress:
![image](https://user-images.githubusercontent.com/80678586/216350044-9dd17f0a-e885-40c1-be3c-c198370fa07c.png)
Could a checkbox like this be helpful? The student can click the checkbox to mark the exercise as done (yellow/blue checkmark), maybe the teacher can see this in a nice overview on the classpage (like your whiteboard), and then check it themself (check turns red or green).

Again, might be hard to work it into the website, idk.

**Reply by Felienne** [reply]
> Another hersenspinsel about helping teachers monitor progress: ![image](https://user-images.githubusercontent.com/80678586/216350044-9dd17f0a-e885-40c1-be3c-c198370fa07c.png) Could a checkbox like this be helpful? The student can click the checkbox to mark the exercise as done (yellow/blue checkmark), maybe the teacher can see this in a nice overview on the classpage (like your whiteboard), and then check it themself (check turns red or green).
> 
> Again, might be hard to work it into the website, idk.

Ow that is pretty cool and simple for the students to use. You can also use it when walking around in class!

**Comment by Felienne** [comment]
Ow what a great idea!!! We can totally do that
